src/facade/dragonfly_connection.cc

```diff
@@ -1333,4 +1343,26 @@ Connection::MemoryUsage Connection::GetMemoryUsage() const {
   };
 }
 
+Connection::WeakRef::WeakRef(std::shared_ptr<Connection> ptr, QueueBackpressure* backpressure,
+                             unsigned thread)
+    : ptr_{ptr}, backpressure_{backpressure}, thread_{thread} {
+}
+
+unsigned Connection::WeakRef::Thread() const {
+  return thread_;
+}
+
+Connection* Connection::WeakRef::Get() const {
+  DCHECK_EQ(ProactorBase::me()->GetPoolIndex(), int(thread_));
+  return ptr_.lock().get();
+}
+
+bool Connection::WeakRef::EnsureMemoryBudget() const {
+  if (!ptr_.expired()) {
```

Comment:
I think that the proper way to go about this is to create a `shared_ptr` out of the `weak_ptr` via `lock()`, and then check if the `shared_ptr` is not null.
However I guess that it's not needed here because we're counting on the fact that this is the same thread as the Fiber that really owns `self_`, and therefore it can't go out of scope in the meanwhile.
Am I right? If so, lgtm.

Reply:
No, we don't rely on it. And it can go out of scope. This code here just checks a threads memory budget

We call `backpressure_->EnsureBelowLimit();`  Backpressure is a pointer to a threadlocal that is never invalidated and we potentially suspend in `EnsureBelowLimit` anyways

So it's more of an optimization: skip checking memory for connection threads that have no active connections placed. If it goes out of scope we just did unnecessary work

Reply:
worth adding a comment about it.

Reply:
Done